models/hash_validator.rb

```diff
@@ -0,0 +1,273 @@
+# Copyright © Mapotempo, 2021
+#
+# This file is part of Mapotempo.
+#
+# Mapotempo is free software. You can redistribute it and/or
+# modify since you respect the terms of the GNU Affero General
+# Public License as published by the Free Software Foundation,
+# either version 3 of the License, or (at your option) any later version.
+#
+# Mapotempo is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+# or FITNESS FOR A PARTICULAR PURPOSE.  See the Licenses for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Mapotempo. If not, see:
+# <http://www.gnu.org/licenses/agpl.html>
+#
+
+require './models/vrp'
+
+module Models
+  class HashValidator < Vrp
+    def initialize(hash, internal_call)
+      @hash = hash
+      @internal_call = internal_call
```

Comment:
I don't think we should have a different behaviour for internal_calls. It is dangerous, we might be able to create invalid VRPs in the code if the validators is not the same.

Reply:
That is right. But then we need to edit _max_split_ behaviour and/or validation regarding number of visits :

- In vrp.create we should reject VRP if service has more than one visit but no schedule. Indeed, if we have no schedule then we do not expand service visits so user will have an inconsistent result.
- In new _max_split_ we edit number of services visits (if I remember well it is in the case were several services have same lat/lon then we generate one service with several visits) but we have no schedule. 

We need to find a way to use _visits_number_ in a more generic way to be able to deal with both cases.

Reply:
- Oh, I thought if we have no `:schedule`, we just ignore the visits_number. If that is the case I don't see how the new split_solve worked in the first place :confused: 
- I think I understand where you mean. Do you mean [here](https://github.com/Mapotempo/optimizer-api/blob/8ae741327b2a4559eecde0f1bef9f607929a3be5/lib/interpreters/split_clustering.rb#L349) where we represent all services coming from the same vehicle "group" with a single point ? This code used to work with the old validators even though we didn't have a separate logic for internal calls, did we change the validators?

In any case, it looks like my usage of `:visits_number` is not completely "appropriate" and I am not happy that we modify the validators because of it.

I can only think of modifying the split_solve so that it first creates the representative_vrp and then sets the visits_number for the visits (which is still an inappropriate usage but at least we don't need to modify the validators because of it). Can you think of another way to avoid this?

Reply:
You are right, this check (there can not be more than one visit without a schedule) is new, that is why it used to work. I guess I added this check in parallel of your changes of split_solve, that is why we never saw the problem. 

As you said, visits_number used to be ignored if no schedule was provided. Maybe it is okay to keep it like that. Documentation about _visits_number_ specifies `Scheduling only`. This check could be helpful for users but no one ever complained about this. Maybe the easier would be to simply remove the check ?

Reply:
Make sense :+1: 

I don't know. In that case we ignore a parameter and the user might not aware of that they made an error :/ or  (I doubt it but) one day we might want to have a "meaning" for `visits_number sans scheduling` and the behaviour would change for json with visits_number without scheduling. I think it's better if you keep the check and remove the internal_call logic and I will fix the split_solve calls (your PR will have a couple of broken tests but I think we can live with that for few days :crossed_fingers:)